src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -305,45 +324,84 @@ private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byt
+                return -1;
+            }
+            else
             {
-                RejectRequestHeader(headerSpan);
+                if (_disableHttp1LineFeedTerminators)
+                {
+                    // LF only but disabled
+
+                    // Advance 1 to include LF in result
+                    lineEnd = currentSlice.GetPosition(1, lineEndPosition.Value);
+                    RejectRequestHeader(currentSlice.Slice(reader.Position, lineEnd).ToSpan());
+                }
             }
+        }
+
+        // At this point the first EOL char is not the last byte in the current slice
+
+        // Offset 1 to include the first EOL char.
+        firstLineEndCharPos = currentSlice.GetPosition(1, lineEndPosition.Value);
+
+        if (firstEolChar == ByteCR)
+        {
+            // First EOL char is CR, include the char after CR
+            lineEnd = currentSlice.GetPosition(2, lineEndPosition.Value);
+            header = currentSlice.Slice(reader.Position, lineEnd);
+        }
+        else if (_disableHttp1LineFeedTerminators)
+        {
+            // The terminator is an LF and we don't allow it.
+            RejectRequestHeader(currentSlice.Slice(reader.Position, firstLineEndCharPos).ToSpan());
             return -1;
         }
+        else
+        {
+            // First EOL char is LF. only include this one
+            lineEnd = currentSlice.GetPosition(1, lineEndPosition.Value);
+            header = currentSlice.Slice(reader.Position, lineEnd);
+        }
 
-        // Advance 2 to include CR{LF?} in lineEnd
-        lineEnd = currentSlice.GetPosition(2, lineEndPosition.Value);
-        headerSpan = currentSlice.Slice(reader.Position, lineEnd).ToSpan();
+        var headerSpan = header.ToSpan();
 
-        if (headerSpan.Length < 5)
+        // 'a:b\n' or 'a:b\r\n'
+        var minHeaderSpan = _disableHttp1LineFeedTerminators ? 5 : 4;
+        if (headerSpan.Length < minHeaderSpan)
         {
-            // Less than min possible headerSpan is 5 bytes a:b\r\n
             RejectRequestHeader(headerSpan);
         }
 
-        if (headerSpan[^2] != ByteCR)
+        var terminatorSize = -1;
+
+        if (headerSpan[^1] == ByteLF)
         {
-            // Sequence needs to be CRLF not LF first.
-            RejectRequestHeader(headerSpan[..^1]);
+            if (headerSpan[^2] == ByteCR)
+            {
+                terminatorSize = 2;
+            }
+            else if (!_disableHttp1LineFeedTerminators)
+            {
+                terminatorSize = 1;
+            }
         }
 
-        if (headerSpan[^1] != ByteLF ||
-            // Exclude the CRLF from the headerLine and parse the header name:value pair
-            !TryTakeSingleHeader(handler, headerSpan[..^2]))
+        // Last chance to bail if the terminator size is not valid or the header doesn't parse.
+        if (terminatorSize == -1 || !TryTakeSingleHeader(handler, headerSpan[..^terminatorSize]))
```

Comment:
I [love slices](https://github.com/dotnet/runtime/issues/66773#issuecomment-1220149348), I think this was one of the bad "const" slice cases [we discussed earlier](https://github.com/dotnet/aspnetcore/pull/43202/files#r945029114) where Roslyn generates better IL using the old syntax:

```suggestion
        if (terminatorSize == -1 || !TryTakeSingleHeader(handler, headerSpan.Slice(0, headerSpan.Length - terminatorSize)))
```

https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQ64GYAEicDC6A3ujuXrgMoAOYAdgDwBGAnsAKYB8OVANgEsAxh1oMA8nwAmACjFM2nHhDr0ANDgH1gOTgCcoWsMAD2eqgIBeHAJRkKpNBWd4A7DhUMAdP2EcZAAwanvReADIc9ADmwAAWOAC0uhwGRqbmVrYA3PbkAL65OIXOhdi8qizs3LyCIvIAchwA7nIVitUhGlo6+ob0xmYW1nZODsUUcO4hANoBXl4Aer1pg5kAujmj+eh5QA